### ISSUE TITLE
Installing Local Dependencies From Monorepo

### ISSUE THREAD

**Issue by harrymconner** [pdm-project/pdm#2527]
I'm moving my discussion post to an issue because, in my opinion, I think that the lock file should allow me to install my project anywhere as long as I have the proper directory structure in place. Happy to be corrected if this is not desirable behavior though!

### Discussed in https://github.com/pdm-project/pdm/discussions/2517

<div type='discussions-op-text'>

<sup>Originally posted by **harrymconner** December 21, 2023</sup>
### Background
I'm having issues deploying a project from within a monorepo structure, and I am trying to determine if this is not possible with PDM or if I am missing something.

The repo is set up as follows:

```
.
├── libraries
│   ├── calculations
│   └── date-manager
└── projects
    └── dashboard
```
### Problem
I'm trying to deploy the "dashboard" project, and on my local machine, all libraries are able to be installed as expected. I start running into issues once I try to install the project anywhere else though because of the absolute paths in the lockfile.

The dependencies for "dashboard" are specified in its pyproject.toml file as below.
```
dependencies = [
    "calculations @ file:///${PROJECT_ROOT}/../../libraries/calculations",
]
```
"dashboard" depends on "calculations", and "calculations", in turn, depends on "date-manager". The "date-manager" transitive dependency is the one that is giving me issues.

Below is what is in "dashboard's" lock file for the "calculations" library. I'm able to install this library perfectly fine.
```
[[package]]
name = "calculations"
version = "0.1.0"
requires_python = ">=3.11"
path = "../../libraries/calculations"
summary = ""
groups = ["default"]
dependencies = [
    "date-manager @ file:///{MY_LOCAL_DIRECTORY}/libraries/calculations/../date-manager",
]
```

I have not been able to install the date-manager library because of the absolute path in the lock file below.
```
[[package]]
name = "date-manager"
version = "0.1.0"
requires_python = ">=3.11"
path = "{MY_LOCAL_DIRECTORY}/libraries/calculations/../date-manager" <-- this path is the source of the issue
summary = ""
groups = ["default"]
dependencies = [
    "python-dateutil>=2.8.2",
]
```

Everything works as expected if I convert the absolute path to a relative path ("../../libraries/date-manager") though. 

### My Question
Is there a way to get pdm to use the relative path in the lock file instead of the absolute path for transitive dependencies?</div>

**Comment by frostming** [comment]
I am not clear how monorepo works in your workflow, but personally I won't keep separate lockfiles for each children project. Instead, I lock all dependencies in the parent project with relative paths and use *named requirements* in children projects to refer to each other.

So in parent `pyproject.toml`, you will have:

```toml
[tool.pdm.dev-dependencies]
workspace = [
    "-e ${PROJECT_ROOT}/libraries/calculations#egg=calculations",
    "-e ${PROJECT_ROOT}/libraries/date-manager#egg=date-manager",
    "-e ${PROJECT_ROOT}/projects/dashboard#egg=dashboard",
 ```

And in `projects/dashboard/pyproject.toml`:

```toml
[project]
dependencies = [
    "calculations"
]
```
`libraries/calculations/pyproject.toml`:

```toml
[project]
dependencies = [
    "date-manager"
]
```

And always run pdm commands in the project root, this will install all children projects in "editable mode", responsive to modifications, and when you publish all of them to the PyPI, you don't need to modify the metadata.

**Comment by harrymconner** [comment]
@frostming Thank you so much for your detailed response! I saw the monorepo example in the docs, but I want to avoid having to install the whole repo into one virtual environment. Your response helped me realize that I need to run all commands from the repo root, and by doing that, I was able to generate a lock file that works for my purposes. Thank you again for your help, and I will close this issue as not planned!
